File: frontend/src/VehicleSilhouettes.js
/* Silhuetas SVG simplificadas mas reconhecíveis para o checklist veicular.
   ViewBox padronizado em 0 0 200 110 — coordenadas dos marks são percentuais
   relativas ao viewBox, então funcionam para qualquer escala (modal e PDF). */
import React from "react";

const STROKE = "#0b1220";
const FILL = "#f1f5f9";

// Side view (lateral) — hatch profile
function SideCar({ flip = false }) {
  return (
    <g transform={flip ? "scale(-1,1) translate(-200,0)" : ""}>
      {/* corpo principal */}
      <path d="M 10 70 L 30 50 L 65 38 L 130 38 L 165 52 L 190 60 L 190 80 L 10 80 Z"
            fill={FILL} stroke={STROKE} strokeWidth="1.5" strokeLinejoin="round" />
      {/* coluna entre janelas */}
      <line x1="98" y1="38" x2="98" y2="60" stroke={STROKE} strokeWidth="1" />
      {/* janela dianteira */}
      <path d="M 38 52 L 65 42 L 95 42 L 95 52 Z"
            fill="#dbeafe" stroke={STROKE} strokeWidth="1" />
      {/* janela traseira */}
      <path d="M 102 42 L 130 42 L 162 52 L 102 52 Z"
            fill="#dbeafe" stroke={STROKE} strokeWidth="1" />
      {/* maçanetas */}
      <line x1="55" y1="62" x2="65" y2="62" stroke={STROKE} strokeWidth="1.2" />
      <line x1="115" y1="62" x2="125" y2="62" stroke={STROKE} strokeWidth="1.2" />
      {/* rodas */}
      <circle cx="50" cy="80" r="11" fill="#0b1220" stroke={STROKE} strokeWidth="1" />
      <circle cx="50" cy="80" r="6" fill="#94a3b8" />
      <circle cx="155" cy="80" r="11" fill="#0b1220" stroke={STROKE} strokeWidth="1" />
      <circle cx="155" cy="80" r="6" fill="#94a3b8" />
      {/* faróis */}
      <ellipse cx="14" cy="68" rx="4" ry="2.5" fill="#fde68a" stroke={STROKE} strokeWidth="0.6" />
      <ellipse cx="186" cy="68" rx="3" ry="2" fill="#fecaca" stroke={STROKE} strokeWidth="0.6" />
    </g>
  );
}

// Front view
function FrontCar() {
  return (
    <>
      {/* corpo */}
      <path d="M 30 90 L 30 50 Q 30 30 60 28 L 140 28 Q 170 30 170 50 L 170 90 Z"
            fill={FILL} stroke={STROKE} strokeWidth="1.5" strokeLinejoin="round" />
      {/* parabrisa */}
      <path d="M 50 50 L 60 32 L 140 32 L 150 50 Z"
            fill="#dbeafe" stroke={STROKE} strokeWidth="1" />
      {/* grade */}
      <rect x="68" y="62" width="64" height="14" fill="#0b1220" stroke={STROKE} strokeWidth="0.8" rx="2" />
      <line x1="72" y1="66" x2="128" y2="66" stroke="#475569" strokeWidth="0.8" />
      <line x1="72" y1="70" x2="128" y2="70" stroke="#475569" strokeWidth="0.8" />
      {/* faróis */}
      <rect x="34" y="55" width="22" height="10" rx="3" fill="#fde68a" stroke={STROKE} strokeWidth="0.8" />
      <rect x="144" y="55" width="22" height="10" rx="3" fill="#fde68a" stroke={STROKE} strokeWidth="0.8" />
      {/* placa */}
      <rect x="84" y="80" width="32" height="8" fill="#fff" stroke={STROKE} strokeWidth="0.6" />
      {/* retrovisores */}
      <rect x="22" y="48" width="8" height="6" fill={FILL} stroke={STROKE} strokeWidth="0.8" rx="2" />
      <rect x="170" y="48" width="8" height="6" fill={FILL} stroke={STROKE} strokeWidth="0.8" rx="2" />
      {/* roda visível */}
      <ellipse cx="44" cy="92" rx="14" ry="3" fill="#0b1220" />
      <ellipse cx="156" cy="92" rx="14" ry="3" fill="#0b1220" />
    </>
  );
}

// Rear view
function RearCar() {
  return (
    <>
      <path d="M 30 90 L 30 50 Q 30 32 60 30 L 140 30 Q 170 32 170 50 L 170 90 Z"
            fill={FILL} stroke={STROKE} strokeWidth="1.5" strokeLinejoin="round" />
      {/* vidro traseiro */}
      <rect x="50" y="38" width="100" height="20" rx="3" fill="#dbeafe" stroke={STROKE} strokeWidth="1" />
      {/* lanternas */}
      <rect x="34" y="62" width="26" height="12" rx="2" fill="#dc2626" stroke={STROKE} strokeWidth="0.8" />
      <rect x="140" y="62" width="26" height="12" rx="2" fill="#dc2626" stroke={STROKE} strokeWidth="0.8" />
      <line x1="35" y1="68" x2="59" y2="68" stroke="#fef2f2" strokeWidth="0.6" />
      <line x1="141" y1="68" x2="165" y2="68" stroke="#fef2f2" strokeWidth="0.6" />
      {/* placa */}
      <rect x="74" y="78" width="52" height="10" rx="1" fill="#fff" stroke={STROKE} strokeWidth="0.6" />
      {/* maçaneta porta-malas */}
      <rect x="92" y="68" width="16" height="3" rx="1" fill="#94a3b8" stroke={STROKE} strokeWidth="0.5" />
      {/* rodas */}
      <ellipse cx="44" cy="92" rx="14" ry="3" fill="#0b1220" />
      <ellipse cx="156" cy="92" rx="14" ry="3" fill="#0b1220" />
    </>
  );
}

// Top view
function TopCar() {
  return (
    <>
      <rect x="40" y="14" width="120" height="82" rx="20" ry="14"
            fill={FILL} stroke={STROKE} strokeWidth="1.5" />
      {/* parabrisas dianteiro */}
      <path d="M 56 30 L 144 30 L 138 44 L 62 44 Z"
            fill="#dbeafe" stroke={STROKE} strokeWidth="1" />
      {/* teto */}
      <rect x="62" y="44" width="76" height="26" fill="#e2e8f0" stroke={STROKE} strokeWidth="0.6" />
      {/* parabrisas traseiro */}
      <path d="M 62 70 L 138 70 L 144 84 L 56 84 Z"
            fill="#dbeafe" stroke={STROKE} strokeWidth="1" />
      {/* divisão portas */}
      <line x1="100" y1="44" x2="100" y2="70" stroke={STROKE} strokeWidth="0.6" strokeDasharray="2 2" />
      {/* retrovisores externos */}
      <rect x="34" y="40" width="6" height="10" fill={FILL} stroke={STROKE} strokeWidth="0.8" rx="1" />
      <rect x="160" y="40" width="6" height="10" fill={FILL} stroke={STROKE} strokeWidth="0.8" rx="1" />
    </>
  );
}

const VIEWS = {
  front:  { label: "Frente",            render: <FrontCar /> },
  rear:   { label: "Traseira",          render: <RearCar /> },
  left:   { label: "Lateral esquerda",  render: <SideCar /> },
  right:  { label: "Lateral direita",   render: <SideCar flip /> },
  top:    { label: "Vista superior",    render: <TopCar /> },
};

export const VIEW_KEYS = Object.keys(VIEWS);
export const VIEW_LABELS = Object.fromEntries(
  Object.entries(VIEWS).map(([k, v]) => [k, v.label])
);

export const DAMAGE_TYPES = {
  D: { label: "Amassado",  color: "#dc2626" },
  S: { label: "Risco",     color: "#d97706" },
  R: { label: "Oxidação",  color: "#7c2d12" },
  F: { label: "Quebrado",  color: "#b91c1c" },
  V: { label: "Vidro",     color: "#0369a1" },
  P: { label: "Pintura",   color: "#475569" },
};

export default function VehicleSilhouette({ view, marks = [], onAddMark, readOnly = false, height = 130 }) {
  const conf = VIEWS[view];
  if (!conf) return null;

  const handleClick = (e) => {
    if (readOnly || !onAddMark) return;
    const svg = e.currentTarget;
    const rect = svg.getBoundingClientRect();
    const x = ((e.clientX - rect.left) / rect.width) * 200;
    const y = ((e.clientY - rect.top) / rect.height) * 110;
    onAddMark({ x: +x.toFixed(1), y: +y.toFixed(1), view });
  };

  return (
    <svg
      viewBox="0 0 200 110"
      width="100%"
      height={height}
      onClick={handleClick}
      style={{
        background: "#fff",
        border: "1px solid var(--border-default)",
        borderRadius: 8,
        cursor: readOnly ? "default" : "crosshair",
        userSelect: "none",
      }}
      data-testid={`vehicle-silhouette-${view}`}
    >
      {conf.render}
      {marks.filter((m) => m.view === view).map((m, i) => {
        const ord = m.ord ?? (i + 1);
        const color = DAMAGE_TYPES[m.code]?.color || "#dc2626";
        return (
          <g key={`${view}-${i}`} pointerEvents="none">
            <circle cx={m.x} cy={m.y} r="6" fill={color} stroke="#fff" strokeWidth="1.5" />
            <text x={m.x} y={m.y + 2.5} fill="#fff" fontSize="7" textAnchor="middle" fontWeight="700">{ord}</text>
          </g>
        );
      })}
    </svg>
  );
}

File: backend/routes/vehicle_silhouettes.py
"""Renderiza as 5 silhuetas do veículo (frente / traseira / lateral esq /
lateral dir / vista superior) no PDF do checklist veicular usando primitives
ReportLab. Coordenadas alinhadas ao SVG do frontend (viewBox 0..200 × 0..110).

Cada silhueta ocupa um Drawing de 200 × 110 px (escala automática do reportlab.platypus).
Marks de avaria são círculos vermelhos numerados (mesmo `ord` do frontend).
"""
from __future__ import annotations


NERVOUS_METADATA = {
    "owner": "ops-team",
    "domain": "operacoes",
    "criticality": "high",
    "emits_events": False,
    "event_types": [],
    "company_id_required": True,
}

from reportlab.graphics.shapes import (Circle, Drawing, Ellipse, Line, Path,
                                        Polygon, Rect, String, definePath)
from reportlab.lib import colors

# Paleta alinhada ao frontend
SLATE = colors.HexColor("#0b1220")
FILL = colors.HexColor("#f1f5f9")
GLASS = colors.HexColor("#dbeafe")
HEADLIGHT = colors.HexColor("#fde68a")
TAILLIGHT = colors.HexColor("#dc2626")
GRILL = colors.HexColor("#0b1220")
TIRE = colors.HexColor("#0b1220")

DAMAGE_COLORS = {
    "D": colors.HexColor("#dc2626"),  # amassado
    "S": colors.HexColor("#d97706"),  # risco
    "R": colors.HexColor("#7c2d12"),  # oxidação
    "F": colors.HexColor("#b91c1c"),  # quebrado
    "V": colors.HexColor("#0369a1"),  # vidro
    "P": colors.HexColor("#475569"),  # pintura
}

VIEW_LABELS = {
    "front":  "Frente",
    "rear":   "Traseira",
    "left":   "Lateral esquerda",
    "right":  "Lateral direita",
    "top":    "Vista superior",
}

# Coords do viewBox SVG: y cresce para baixo (web). ReportLab cresce para
# cima. Precisamos espelhar y → y_rl = 110 - y_svg em desenhos.


def _flip_y(y: float, h: float = 110.0) -> float:
    return h - y


def _path_filled(d: Drawing, points: list, fill, stroke=SLATE, sw=1.2):
    """Polígono fechado com lista [(x,y), ...] em coord SVG (y cresce p/ baixo)."""
    pts = []
    for x, y in points:
        pts.extend([x, _flip_y(y)])
    d.add(Polygon(points=pts, fillColor=fill, strokeColor=stroke, strokeWidth=sw))


def _line(d: Drawing, x1, y1, x2, y2, stroke=SLATE, sw=1.0, dash=None):
    ln = Line(x1, _flip_y(y1), x2, _flip_y(y2), strokeColor=stroke, strokeWidth=sw)
    if dash:
        ln.strokeDashArray = list(dash) if isinstance(dash, (tuple, list)) else [2, 2]
    d.add(ln)


def _rect(d: Drawing, x, y, w, h, fill, stroke=SLATE, sw=1.0):
    """Retângulo em coord SVG (y cresce p/ baixo). y é o topo."""
    d.add(Rect(x, _flip_y(y + h), w, h, fillColor=fill, strokeColor=stroke, strokeWidth=sw))


def _circle(d: Drawing, cx, cy, r, fill, stroke=SLATE, sw=1.0):
    d.add(Circle(cx, _flip_y(cy), r, fillColor=fill, strokeColor=stroke, strokeWidth=sw))


def _ellipse(d: Drawing, cx, cy, rx, ry, fill, stroke=SLATE, sw=1.0):
    d.add(Ellipse(cx, _flip_y(cy), rx, ry, fillColor=fill, strokeColor=stroke, strokeWidth=sw))


def _text(d: Drawing, x, y, txt, color=colors.white, size=7, bold=True):
    s = String(x, _flip_y(y) - size / 3, txt, fillColor=color, fontSize=size, textAnchor="middle")
    if bold:
        s.fontName = "Helvetica-Bold"
    d.add(s)


# ---------------------------------------------------------------------------
# Vistas
# ---------------------------------------------------------------------------
def _draw_side(d: Drawing, flip: bool = False):
    # Quando flip=True (lado direito), espelha em x.
    def fx(x: float) -> float:
        return 200 - x if flip else x

    # Corpo
    body = [(fx(10), 70), (fx(30), 50), (fx(65), 38), (fx(130), 38),
            (fx(165), 52), (fx(190), 60), (fx(190), 80), (fx(10), 80)]
    _path_filled(d, body, FILL, SLATE, 1.2)
    # Coluna entre janelas
    _line(d, fx(98), 38, fx(98), 60, SLATE, 0.8)
    # Janela dianteira
    _path_filled(d, [(fx(38), 52), (fx(65), 42), (fx(95), 42), (fx(95), 52)], GLASS, SLATE, 0.8)
    # Janela traseira
    _path_filled(d, [(fx(102), 42), (fx(130), 42), (fx(162), 52), (fx(102), 52)], GLASS, SLATE, 0.8)
    # Maçanetas
    _line(d, fx(55), 62, fx(65), 62, SLATE, 1.0)
    _line(d, fx(115), 62, fx(125), 62, SLATE, 1.0)
    # Rodas
    _circle(d, fx(50), 80, 11, TIRE, SLATE, 0.8)
    _circle(d, fx(50), 80, 6, colors.HexColor("#94a3b8"), SLATE, 0.4)
    _circle(d, fx(155), 80, 11, TIRE, SLATE, 0.8)
    _circle(d, fx(155), 80, 6, colors.HexColor("#94a3b8"), SLATE, 0.4)
    # Faróis
    _ellipse(d, fx(14), 68, 4, 2.5, HEADLIGHT, SLATE, 0.5)
    _ellipse(d, fx(186), 68, 3, 2, colors.HexColor("#fecaca"), SLATE, 0.5)


def _draw_front(d: Drawing):
    # Corpo (cantos arredondados aproximados via polígono)
    body = [(30, 90), (30, 50), (45, 32), (60, 28), (140, 28), (155, 32),
            (170, 50), (170, 90)]
    _path_filled(d, body, FILL, SLATE, 1.2)
    # Parabrisa
    _path_filled(d, [(50, 50), (60, 32), (140, 32), (150, 50)], GLASS, SLATE, 0.8)
    # Grade
    _rect(d, 68, 62, 64, 14, GRILL, SLATE, 0.6)
    _line(d, 72, 66, 128, 66, colors.HexColor("#475569"), 0.6)
    _line(d, 72, 70, 128, 70, colors.HexColor("#475569"), 0.6)
    # Faróis
    _rect(d, 34, 55, 22, 10, HEADLIGHT, SLATE, 0.6)
    _rect(d, 144, 55, 22, 10, HEADLIGHT, SLATE, 0.6)
    # Placa
    _rect(d, 84, 80, 32, 8, colors.white, SLATE, 0.5)
    # Retrovisores
    _rect(d, 22, 48, 8, 6, FILL, SLATE, 0.6)
    _rect(d, 170, 48, 8, 6, FILL, SLATE, 0.6)
    # Sombra das rodas
    _ellipse(d, 44, 92, 14, 3, TIRE, SLATE, 0.4)
    _ellipse(d, 156, 92, 14, 3, TIRE, SLATE, 0.4)


def _draw_rear(d: Drawing):
    body = [(30, 90), (30, 50), (45, 34), (60, 30), (140, 30), (155, 34),
            (170, 50), (170, 90)]
    _path_filled(d, body, FILL, SLATE, 1.2)
    # Vidro traseiro
    _rect(d, 50, 38, 100, 20, GLASS, SLATE, 0.8)
    # Lanternas
    _rect(d, 34, 62, 26, 12, TAILLIGHT, SLATE, 0.6)
    _rect(d, 140, 62, 26, 12, TAILLIGHT, SLATE, 0.6)
    _line(d, 35, 68, 59, 68, colors.HexColor("#fef2f2"), 0.5)
    _line(d, 141, 68, 165, 68, colors.HexColor("#fef2f2"), 0.5)
    # Placa
    _rect(d, 74, 78, 52, 10, colors.white, SLATE, 0.5)
    # Maçaneta porta-malas
    _rect(d, 92, 68, 16, 3, colors.HexColor("#94a3b8"), SLATE, 0.4)
    # Sombra rodas
    _ellipse(d, 44, 92, 14, 3, TIRE, SLATE, 0.4)
    _ellipse(d, 156, 92, 14, 3, TIRE, SLATE, 0.4)


def _draw_top(d: Drawing):
    # Corpo (aproximação de retângulo arredondado via polígono)
    body = [(40, 32), (60, 14), (140, 14), (160, 32),
            (160, 78), (140, 96), (60, 96), (40, 78)]
    _path_filled(d, body, FILL, SLATE, 1.2)
    # Parabrisa dianteiro
    _path_filled(d, [(56, 30), (144, 30), (138, 44), (62, 44)], GLASS, SLATE, 0.8)
    # Teto
    _rect(d, 62, 44, 76, 26, colors.HexColor("#e2e8f0"), SLATE, 0.5)
    # Parabrisa traseiro
    _path_filled(d, [(62, 70), (138, 70), (144, 84), (56, 84)], GLASS, SLATE, 0.8)
    # Divisão de portas
    _line(d, 100, 44, 100, 70, SLATE, 0.5, [2, 2])
    # Retrovisores
    _rect(d, 34, 40, 6, 10, FILL, SLATE, 0.6)
    _rect(d, 160, 40, 6, 10, FILL, SLATE, 0.6)


_VIEW_DRAWERS = {
    "front": _draw_front,
    "rear": _draw_rear,
    "top": _draw_top,
}


def silhouette_drawing(view: str, marks: list, width: float, height: float) -> Drawing:
    """Cria um Drawing pronto pra inserir no Story do platypus.

    width/height em pontos (1cm = 28.346 pt). O viewBox interno (200×110) é
    escalado proporcionalmente.
    """
    # Drawing usa coords de impressão (y cresce para cima), mas nossos
    # _flip_y já convertem do SVG.
    d = Drawing(200, 110)
    if view == "left":
        _draw_side(d, flip=False)
    elif view == "right":
        _draw_side(d, flip=True)
    else:
        drawer = _VIEW_DRAWERS.get(view)
        if drawer:
            drawer(d)
    # Marks
    for m in marks:
        if m.get("view") != view:
            continue
        color = DAMAGE_COLORS.get(m.get("code", "D"), DAMAGE_COLORS["D"])
        cx, cy = float(m.get("x", 0)), float(m.get("y", 0))
        _circle(d, cx, cy, 6, color, colors.white, 1.5)
        _text(d, cx, cy, str(m.get("ord", "?")), color=colors.white, size=7)
    # Escala final
    d.width = width
    d.height = height
    d.scale(width / 200.0, height / 110.0)
    return d
